wsi/drm/drmdisplay.cpp

```diff
@@ -419,30 +408,81 @@ bool DrmDisplay::SetBroadcastRGB(const char *range_property) {
   return true;
 }
 
-void DrmDisplay::SetHDCPState(HWCContentProtection state,
+bool DrmDisplay::CheckHDCPStatus() {
+  int val = -1;
+  uint32_t i;
+  ScopedDrmObjectPropertyPtr cprops(drmModeObjectGetProperties(
+      gpu_fd_, connector_, DRM_MODE_OBJECT_CONNECTOR));
+  for (i = 0; i < cprops->count_props; i++) {
+    if (cprops->props[i] == hdcp_id_prop_) {
+      val = cprops->prop_values[i];
+      break;
+    }
+  }
+
+  if ((current_protection_support_ && val == HWCContentProtection::kEnabled) ||
+      (!current_protection_support_ && val == HWCContentProtection::kDesired))
+    return true;
+
+  return false;
+}
+
+bool DrmDisplay::SetHDCPState(HWCContentProtection state,
                               HWCContentType content_type) {
   desired_protection_support_ = state;
-  content_type_ = content_type;
-  if (desired_protection_support_ == current_protection_support_)
-    return;
+  desired_content_type_ = content_type;
+
+  if (desired_protection_support_ == current_protection_support_ &&
+      desired_content_type_ == current_content_type_)
+    return true;
 
   if (hdcp_id_prop_ <= 0) {
     ETRACE("Cannot set HDCP state as Connector property is not supported \n");
-    return;
+    return false;
+  }
+  if ((hdcp_type_prop_ <= 0) && (desired_content_type_ == 1)) {
+    ETRACE("Content type property unavailable. HDCP 2.2 cannot be enabled\n");
+    return false;
   }
 
   if (!(connection_state_ & kConnected)) {
-    return;
+    return false;
   }
 
-  current_protection_support_ = desired_protection_support_;
-  uint32_t value = 0;
-  if (current_protection_support_ == kDesired) {
-    value = 1;
+  ScopedDrmAtomicReqPtr pset(drmModeAtomicAlloc());
+  int ret = drmModeAtomicAddProperty(pset.get(), connector_, hdcp_id_prop_,
+                                     desired_protection_support_) < 0;
+  if (current_protection_support_ && (hdcp_type_prop_ > 0))
+    ret = ret ||
+          drmModeAtomicAddProperty(pset.get(), connector_, hdcp_type_prop_,
+                                   desired_content_type_) < 0;
+  if (ret) {
+    ETRACE("Unable to fill pset for HDCP\n");
+    return false;
+  }
+  ret = drmModeAtomicCommit(gpu_fd_, pset.get(), DRM_MODE_ATOMIC_ALLOW_MODESET,
+                            NULL);
+  if (ret) {
+    ETRACE("Failed to commit HDCP ret %d errno %d : %s\n", ret, errno,
+           PRINTERROR());
+    return false;
   }
 
-  drmModeConnectorSetProperty(gpu_fd_, connector_, hdcp_id_prop_, value);
-  ETRACE("Ignored Content type. \n");
+  /* Wait for Enable should be 5.1 Sec * 3(kernel reattempt cnt) */
```

Comment:
possible to wait up to 15s  blocking hwc service such long time seems not acceptable. I would suggest not blocking in hwc.

Reply:
Yes, agreed. This should be done by the hdcp daemon. I had previously raised it as a concern.
Once the daemon does it, We can remove this here.